src/functions.php

```diff
@@ -42,6 +43,39 @@
 use function reset;
 use function substr;
 
+/**
+ * Check whether all servers support executing a write stage on a secondary.
+ *
+ * @internal
+ * @see https://github.com/mongodb/specifications/blob/master/source/crud/crud.rst#aggregation-pipelines-with-write-stages
+ */
+function all_servers_support_write_stage_on_secondary(Manager $manager): bool
+{
+    static $wireVersionForWriteStageOnSecondary = 13;
+
+    $servers = $manager->getServers();
+
+    /* If no servers are available, assume that SDAM has not been initialized
+     * and make a single server selection attempt to do so now. We do not
+     * actually care about selecting a server so any error can be ignored. */
+    if (empty($servers)) {
+        try {
+            $manager->selectServer(new ReadPreference(ReadPreference::NEAREST));
```

Comment:
IIUC the intent is here to initialize SDAM and do one full scan. `ReadPreference::NEAREST` is arbitrary.

Could this use the user's read preference, and return the selected matching server on success?

I think that could avoid this (rare) scenario:

- MongoDB cluster is: P5 S5 S4
- ReadPreference=Secondary and read preference tags are set to only match S4.
- S4 is unavailable initially. The initial scan keeps S4 as Unknown.
- The S4 becomes available shortly after.
- The next server selection selects waits for matching read preference tags and selects the S4.

Reply:
One concern would be if server selection fails and then we realize that all servers are 5.0+. In that case, the exception we ignored is actually what should be thrown. That said, I can just save the exception and rethrow it later in the function.

My original idea was to create a separate `select_server()` function specifically for `$out` and `$merge`, but I backpedaled to this boolean function to keep things simpler. I can try to rework this function by reverting it from a boolean check back to more complex server selection function.